pkg/pillar/cmd/downloader/context.go

```diff
@@ -0,0 +1,181 @@
+	ctx.subGlobalDownloadConfig = subGlobalDownloadConfig
+	subGlobalDownloadConfig.Activate()
+
+	// Look for DatastoreConfig. We should process this
+	// before any download config ( App/baseos/cert). Without DataStore Config,
+	// Image Downloads will run into errors.
+	subDatastoreConfig, err := pubsub.Subscribe("zedagent",
+		types.DatastoreConfig{}, false, &ctx)
+	if err != nil {
+		return err
+	}
+	subDatastoreConfig.ModifyHandler = handleDatastoreConfigModify
+	subDatastoreConfig.CreateHandler = handleDatastoreConfigModify
+	subDatastoreConfig.DeleteHandler = handleDatastoreConfigDelete
+	ctx.subDatastoreConfig = subDatastoreConfig
+	subDatastoreConfig.Activate()
+
+	pubGlobalDownloadStatus, err := pubsub.Publish(agentName,
+		types.GlobalDownloadStatus{})
+	if err != nil {
+		return err
+	}
+	ctx.pubGlobalDownloadStatus = pubGlobalDownloadStatus
+
+	// Set up our publications before the subscriptions so ctx is set
+	pubAppImgStatus, err := pubsub.PublishScope(agentName, types.AppImgObj,
+		types.DownloaderStatus{})
+	if err != nil {
+		return err
+	}
+	ctx.pubAppImgStatus = pubAppImgStatus
+	pubAppImgStatus.ClearRestarted()
+
+	pubBaseOsStatus, err := pubsub.PublishScope(agentName, types.BaseOsObj,
+		types.DownloaderStatus{})
+	if err != nil {
+		return err
+	}
+	ctx.pubBaseOsStatus = pubBaseOsStatus
+	pubBaseOsStatus.ClearRestarted()
+
+	pubCertObjStatus, err := pubsub.PublishScope(agentName, types.CertObj,
+		types.DownloaderStatus{})
+	if err != nil {
+		return err
+	}
+	ctx.pubCertObjStatus = pubCertObjStatus
+	pubCertObjStatus.ClearRestarted()
+
+	subAppImgConfig, err := pubsub.SubscribeScope("zedmanager",
+		types.AppImgObj, types.DownloaderConfig{}, false, &ctx)
+	if err != nil {
+		return err
+	}
+	subAppImgConfig.ModifyHandler = handleAppImgModify
+	subAppImgConfig.CreateHandler = handleAppImgCreate
+	subAppImgConfig.DeleteHandler = handleAppImgDelete
+	ctx.subAppImgConfig = subAppImgConfig
+	subAppImgConfig.Activate()
+
+	subBaseOsConfig, err := pubsub.SubscribeScope("baseosmgr",
+		types.BaseOsObj, types.DownloaderConfig{}, false, &ctx)
+	if err != nil {
+		return err
+	}
+	subBaseOsConfig.ModifyHandler = handleBaseOsModify
+	subBaseOsConfig.CreateHandler = handleBaseOsCreate
+	subBaseOsConfig.DeleteHandler = handleBaseOsDelete
+	ctx.subBaseOsConfig = subBaseOsConfig
+	subBaseOsConfig.Activate()
+
+	subCertObjConfig, err := pubsub.SubscribeScope("baseosmgr",
+		types.CertObj, types.DownloaderConfig{}, false, &ctx)
+	if err != nil {
+		return err
+	}
+	subCertObjConfig.ModifyHandler = handleCertObjModify
+	subCertObjConfig.CreateHandler = handleCertObjCreate
+	subCertObjConfig.DeleteHandler = handleCertObjDelete
+	ctx.subCertObjConfig = subCertObjConfig
+	subCertObjConfig.Activate()
+
+	pubAppImgStatus.SignalRestarted()
+	pubBaseOsStatus.SignalRestarted()
+	pubCertObjStatus.SignalRestarted()
+
+	return nil
+}
+
+func (ctx *downloaderContext) subscription(objType string) *pubsub.Subscription {
```

Comment:
We generally used get/set convention.. Are you proposing drop "get" ?

Reply:
I am fine with get/set, although it is more of a vestige of Java style than Go style, there is no reason why we cannot do it. In any case, the whole `downloaderContext` and its associated funcs are private and not exported. If I start to see a pattern of usage across other areas outside of downloader, happy to look at standardizing as an interface, though.

Reply:
Ah.. You are right - Lets stick with Go convention... No "Get" for getters - but use "Set" for Setters:

Go doesn't provide automatic support for getters and setters. There's nothing wrong with providing getters and setters yourself, and it's often appropriate to do so, but it's neither idiomatic nor necessary to put Get into the getter's name. If you have a field called owner (lower case, unexported), the getter method should be called Owner (upper case, exported), not GetOwner. The use of upper-case names for export provides the hook to discriminate the field from the method. A setter function, if needed, will likely be called SetOwner. Both names read well in practice:
https://golang.org/doc/effective_go.html#interface-names

Reply:
The bigger issue here is that because it is all private, there are few conventions. If there were private and public properties, that make lean in a certain direction.

On the other hand, most people seem to like `func` names that reflect what it does, e.g. in this case `SubscriptionByType()` might be more reflective of it. And in this case, there really is no setter, as it is set on instantiation.

So, if we see no other way we expect to get a `Subscription`, i.e. the canonical way to get a `Subscription` _is_ by type, then I would leave it as is, i.e. `subscription()`. If, on the other hand, we see there might be lots of ways of retrieving a particular `Subscription` (I don't see how, but you might), then I would rename it to `subscriptionByType()`.

Thoughts @kalyan-nidumolu ?

Reply:
I think as of now, it is by type. So we can leave it as subscription()